libcontainer/container_linux.go

```diff
@@ -454,11 +454,43 @@ func slicesContains[S ~[]E, E comparable](slice S, needle E) bool {
 	return false
 }
 
-func isDmzBinarySafe(c *configs.Config) bool {
+// No longer needed in Go 1.21.
+func slicesEqual[S ~[]E, E comparable](s1, s2 S) bool {
+	if len(s1) != len(s2) {
+		return false
+	}
+	for idx := range s1 {
+		if s1[idx] != s2[idx] {
+			return false
+		}
+	}
+	return true
+}
+
+func shouldUseDmzBinary(p *Process, c *configs.Config) bool {
```

Comment:
Can we just make dmz opt-in via an env or a CLI flag?

- https://github.com/opencontainers/runc/issues/4125#issuecomment-1844220644

Reply:
That would be the simplest solution, but it seems like a bit of a shame to have this code and not use it... Should we remove the SELinux logic too?

Reply:
Probably we should define a ternary env var like `RUNC_USE_DMZ=(1|0|auto)`.

The default value should be `auto`, however, for runc v1.2, I'd suggest to just treat this as an alias for `0` (false) to minimize the incompatibility.

In a future version of runc, we may implement more clever logic for `auto`.

Reply:
cc @lifubang WDYT?

Reply:
`RUNC_DMZ=legacy` can disable the dmz feature now. You mean you worry about there will be more imcompatible reasons not included in #4158 ? But we should know that if we set the default value to `legacy`, the k8s e2e test case about this area will fail? How to improve this test case in k8s?

Reply:
I think we might masquerade this in k8s if we disable runc-dmz if the container is not running as root. I think if it runs as root we don't need to change the capabilities.

I'm not sure if the root detection is hard or not safe and that is why it wasn't done here. I haven't looked into it.

Reply:
Someone mentioned that checking if we are root would be sufficient. To be honest, I struggle to understand all of the interactions of capabilities with everything else in the kernel (some of the functions in commoncap are actual line noise to my eyes).

The issue is that runc binary overwrites are only relevant for uid 0 in most cases. However, if `runc-dmz` is only used for unprivileged container users maybe that'd be okay for now (not uid 0 and no caps).